File: lib/agent/nodes/validate_and_safety.ts
// lib/agent/nodes/validate_and_safety.ts
// replaces sql_validate + safety_check as two separate nodes
import { validateAndCorrectSQL } from '../../sql-validator';
import type { AgentStateType } from '../state';
import type { SchemaTable } from '../../types';
import { logQueryStep } from '../../query-logger';

const MUTATION_PATTERNS = /\b(DELETE|UPDATE|DROP|ALTER|INSERT)\b/i;

function standardizeIdentifiers(
  sql: string,
  dbDialect: string,
  schemaInfo: SchemaTable[] | null
): { correctedSql: string; corrections: string[]; unmatchedIdentifiers: string[] } {
  if (!schemaInfo) {
    return { correctedSql: sql, corrections: [], unmatchedIdentifiers: [] };
  }
  return validateAndCorrectSQL(sql, schemaInfo, dbDialect);
}

export async function validateAndSafetyNode(
  state: AgentStateType
): Promise<Partial<AgentStateType>> {
  const sql = state.generatedSQL;
  if (!sql) return {};

  // --- Part 1: Table/Column Existence Check (from sql_validate) ---
  const validation = validateTablesExistInSchema(sql, state.schemaInfo);

  if (!validation.valid) {
    // Route straight to responder — no retry, LLM already had its chance
    const availableTables = state.schemaInfo
      ?.map(t => `- ${t.name} (${t.columns.map(c => c.name).join(', ')})`)
      .join('\n') ?? '';

    logQueryStep({
      type: 'validation',
      sql: `-- Validation Failed: Unknown tables: ${validation.unknownTables.join(', ')}`,
      success: false,
      error: `Unknown tables: ${validation.unknownTables.join(', ')}`,
    });

    return {
      generatedSQL: "",
      error: `Unknown tables: ${validation.unknownTables.join(', ')}`,
      finalResponse: {
        type: "error",
        message: `The query references tables that don't exist in your database: ${validation.unknownTables.join(', ')}.\n\nAvailable tables:\n${availableTables}`
      }
    };
  }

  // --- Part 2: Identifier Standardization (from sql_validate) ---
  const result = standardizeIdentifiers(sql, state.dbDialect, state.schemaInfo);
  const standardizedSQL = result.correctedSql;

  // --- Part 3: Safety Check (from safety_check) ---
  const isMutation = MUTATION_PATTERNS.test(standardizedSQL);
  const mutationMatch = standardizedSQL.match(MUTATION_PATTERNS);
  const mutationType = mutationMatch ? mutationMatch[1].toUpperCase() : '';

  const update: Partial<AgentStateType> = {
    generatedSQL: standardizedSQL,
    isMutation,
    mutationType,
    identifierCorrections: result.corrections,
    unmatchedIdentifiers: result.unmatchedIdentifiers,
  };

  logQueryStep({
    type: 'validation',
    sql: `-- Validation Check: Standardizing identifiers & safety check\n-- Corrected SQL Casing/Quotes: ${standardizedSQL}${result.corrections.length > 0 ? `\n-- Corrections applied: ${result.corrections.join(', ')}` : ''}`,
    success: true,
  });

  if (state.finalResponse) {
    update.finalResponse = {
      ...state.finalResponse,
      sql: standardizedSQL,
      identifierCorrections: result.corrections,
      unmatchedIdentifiers: result.unmatchedIdentifiers,
      isMutation,
      mutationType,
      pendingApproval: isMutation ? true : undefined,
    };
  }

  return update;
}

function validateTablesExistInSchema(
  sql: string,
  schemaInfo: SchemaTable[] | null
): { valid: boolean; unknownTables: string[] } {
  if (!schemaInfo) return { valid: true, unknownTables: [] };

  const knownTables = schemaInfo.map(t => t.name.toLowerCase());

  // Extract table names from FROM, JOIN, INTO, UPDATE clauses
  const tablePattern = /(?:FROM|JOIN|INTO|UPDATE)\s+[`"']?(\w+)[`"']?/gi;
  const matches = [...sql.matchAll(tablePattern)];
  const usedTables = matches.map(m => m[1].toLowerCase());

  // Exclude CTE names — they're defined in the same query
  const ctePattern = /WITH\s+[`"']?(\w+)[`"']?\s+AS/gi;
  const cteMatches = [...sql.matchAll(ctePattern)];
  const cteNames = cteMatches.map(m => m[1].toLowerCase());

  const unknownTables = usedTables.filter(
    t => !knownTables.includes(t) && !cteNames.includes(t) && t !== 'dual'
  );

  return { valid: unknownTables.length === 0, unknownTables };
}

File: backend/agent/nodes/validate_and_safety.py
import re
from typing import Dict, Any, List, Tuple
from sql_validator import validate_and_correct_sql

MUTATION_PATTERNS = re.compile(r'\b(DELETE|UPDATE|DROP|ALTER|INSERT)\b', re.IGNORECASE)

def validate_tables_exist_in_schema(sql: str, schema_info: List[Dict[str, Any]]) -> Tuple[bool, List[str]]:
    if not schema_info:
        return True, []

    known_tables = {t.get("name", "").lower() for t in schema_info}

    # Extract tables from FROM, JOIN, INTO, UPDATE
    table_pattern = re.compile(r'(?:FROM|JOIN|INTO|UPDATE)\s+[`"\']?(\w+)[`"\']?', re.IGNORECASE)
    used_tables = [m.group(1).lower() for m in table_pattern.finditer(sql)]

    # CTE names
    cte_pattern = re.compile(r'WITH\s+[`"\']?(\w+)[`"\']?\s+AS', re.IGNORECASE)
    cte_names = {m.group(1).lower() for m in cte_pattern.finditer(sql)}

    unknown_tables = [
        t for t in used_tables
        if t not in known_tables and t not in cte_names and t != 'dual'
    ]

    # Deduplicate unknown tables
    unknown_tables = list(set(unknown_tables))

    return len(unknown_tables) == 0, unknown_tables

def validate_and_safety_node(state: Dict[str, Any]) -> Dict[str, Any]:
    sql = state.get("generated_sql")
    if not sql:
        return {}

    schema_info = state.get("schema_info") or []

    # 1. Table Existence Check
    valid, unknown_tables = validate_tables_exist_in_schema(sql, schema_info)
    if not valid:
        available = []
        for t in schema_info:
            cols = ", ".join([col.get("name") for col in t.get("columns", [])])
            available.append(f"- {t.get('name')} ({cols})")
        available_str = "\n".join(available)

        return {
            "generated_sql": "",
            "error": f"Unknown tables: {', '.join(unknown_tables)}",
            "final_response": {
                "type": "error",
                "message": f"The query references tables that don't exist in your database: {', '.join(unknown_tables)}.\n\nAvailable tables:\n{available_str}"
            }
        }

    # 2. Identifier Standardization
    db_dialect = state.get("db_dialect", "sqlite")
    result = validate_and_correct_sql(sql, schema_info, db_dialect)
    standardized_sql = result.get("correctedSql", sql)

    # 3. Safety Check
    is_mutation = bool(MUTATION_PATTERNS.search(standardized_sql))
    mutation_match = MUTATION_PATTERNS.search(standardized_sql)
    mutation_type = mutation_match.group(1).upper() if mutation_match else ""

    update = {
        "generated_sql": standardized_sql,
        "is_mutation": is_mutation,
        "mutation_type": mutation_type,
        "identifier_corrections": result.get("corrections", []),
        "unmatched_identifiers": result.get("unmatchedIdentifiers", [])
    }

    final_resp = state.get("final_response")
    if final_resp:
        # copy dict to avoid changing original state dict by reference
        updated_final = dict(final_resp)
        updated_final.update({
            "sql": standardized_sql,
            "identifierCorrections": result.get("corrections", []),
            "unmatchedIdentifiers": result.get("unmatchedIdentifiers", []),
            "isMutation": is_mutation,
            "mutationType": mutation_type
        })
        if is_mutation:
            updated_final["pendingApproval"] = True
        update["final_response"] = updated_final

    return update
